### PR TITLE
filter_nightfall: Add initial implementation

### PR DESCRIPTION
**Pull request by victor88121**
This patch adds an initial implementation of the Nightfall filter, which can be

used to scan logs for sensitive info like API keys or PII.

Signed-off-by: Victor Chen <victor88121@hotmail.com>

- new filter plugin that uses [Nightfall's](https://docs.nightfall.ai/reference/scanpayloadv3) scanning API to scan for sensitive info and redact the sensitive portions.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [N/A] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [x] Documentation required for this feature

**Backporting**

- [N/A] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by victor88121** [comment]
Example config

```
[FILTER]
    Name nightfall
    Match *
    nightfall_api_key <API key>
    policy_id 5991946b-1cc8-4c38-9240-72677029a3f7
    sampling_rate 1
```

**Comment by victor88121** [comment]
Output

```
bin/fluent-bit -c ~/GolandProjects/filter.conf
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/09 19:46:22] [ info] [engine] started (pid=53844)
[2022/02/09 19:46:22] [ info] [storage] version=1.1.5, initializing...
[2022/02/09 19:46:22] [ info] [storage] in-memory
[2022/02/09 19:46:22] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/09 19:46:22] [ info] [cmetrics] version=0.2.2
[2022/02/09 19:46:22] [ info] [input:http:http.0] listening on 0.0.0.0:8000
[2022/02/09 19:46:22] [ info] [sp] stream processor started
[2022/02/09 19:46:30] [ info] [filter:nightfall:nightfall.0] Nightfall request http_do=0, HTTP Status: 200
[0] app.log: [1644464790.280412000, {"A"=>"there is nothing sensitive here", "B"=>[{"A"=>"my credit card number is *******************"}, {"A"=>"*********** is my social security."}], "C"=>false, "D"=>"key ********************"}]
[2022/02/09 19:47:25] [ info] [filter:nightfall:nightfall.0] Nightfall request http_do=0, HTTP Status: 200
[0] app.log: [1644464845.675431000, {"A"=>"a very safe string"}]
```
Curl
```
curl -XPOST -H "Content-Type: application/json" -d '{"A":"there is nothing sensitive here","B":[{"A":"my credit card number is 4242 4242 4242 4242"},{"A":"539 21 1934 is my social security."}],"C":false,"D":"key kynADfnWAkAV8pkYmtui"}' localhost:8000/app.log
curl -XPOST -H "Content-Type: application/json" -d '{"A":"a very safe string"}' localhost:8000/app.log
```

File: plugins/filter_nightfall/nightfall_api.c [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    new_obj->cur_index = 0;
+    new_obj->start_at_val = FLB_FALSE;
+    mk_list_init(&stack);
+    mk_list_add(&new_obj->_head, &stack);
+
+    mk_list_init(&payload_list);
+
+    /* 
+     * Since logs can contain many levels of nested objects, use stack-based DFS here
+     * to extract scannable fields (positive/negative ints, strings)
```
Comment (victor88121) [inline plugins/filter_nightfall/nightfall_api.c:75]: wasn't sure if there was already a built in function to go through every field in a log so went with this, happy to be pointed to one if there is

**Comment by victor88121** [comment]
valgrind reported leakage in the HTTP input that I used to test which doesn't seem related to this plugin. When I used other inputs (random, etc) there were no leaks.
```
valgrind --leak-check=full bin/fluent-bit -c /filter.conf
==12799== Memcheck, a memory error detector
==12799== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==12799== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==12799== Command: bin/fluent-bit -c /filter.conf
==12799== 
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/14 14:56:35] [ info] [engine] started (pid=12799)
[2022/02/14 14:56:35] [ info] [storage] version=1.1.5, initializing...
[2022/02/14 14:56:35] [ info] [storage] in-memory
[2022/02/14 14:56:35] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/14 14:56:35] [ info] [cmetrics] version=0.2.2
[2022/02/14 14:56:35] [ info] [input:http:http.0] listening on 0.0.0.0:8000
[2022/02/14 14:58:29] [ info] [sp] stream processor started
[2022/02/14 15:00:51] [ info] [filter:nightfall:nightfall.0] Nightfall request http_do=0, HTTP Status: 200
[0] app.log: [1644886849.845300545, {"A"=>"there is nothing sensitive here", "B"=>[{"A"=>"my credit card number is *******************"}, {"A"=>"*********** is my social security."}], "C"=>false, "D"=>"key ********************"}]
^C[2022/02/14 15:01:00] [engine] caught signal (SIGINT)
[2022/02/14 15:01:00] [ warn] [engine] service will shutdown in max 5 seconds
[2022/02/14 15:01:01] [ info] [engine] service has stopped (0 pending tasks)
==12799== 
==12799== HEAP SUMMARY:
==12799==     in use at exit: 56 bytes in 1 blocks
==12799==   total heap usage: 15,530 allocs, 15,529 frees, 9,966,961 bytes allocated
==12799== 
==12799== 56 bytes in 1 blocks are definitely lost in loss record 1 of 1
==12799==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==12799==    by 0x4C396E: mk_mem_alloc (mk_memory.h:53)
==12799==    by 0x4C3BD6: mk_channel_new (mk_stream.c:29)
==12799==    by 0x1F6A72: http_conn_session_init (http_conn.c:174)
==12799==    by 0x1F6DEC: http_conn_add (http_conn.c:247)
==12799==    by 0x1F5A00: in_http_collect (http.c:50)
==12799==    by 0x17C186: flb_input_collector_fd (flb_input.c:1102)
==12799==    by 0x190BCA: flb_engine_handle_event (flb_engine.c:413)
==12799==    by 0x190BCA: flb_engine_start (flb_engine.c:705)
==12799==    by 0x16F579: flb_lib_worker (flb_lib.c:627)
==12799==    by 0x4857608: start_thread (pthread_create.c:477)
==12799==    by 0x4B03292: clone (clone.S:95)
==12799== 
==12799== LEAK SUMMARY:
==12799==    definitely lost: 56 bytes in 1 blocks
==12799==    indirectly lost: 0 bytes in 0 blocks
==12799==      possibly lost: 0 bytes in 0 blocks
==12799==    still reachable: 0 bytes in 0 blocks
==12799==         suppressed: 0 bytes in 0 blocks
==12799== 
==12799== For lists of detected and suppressed errors, rerun with: -s
==12799== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Comment by victor88121** [comment]
@fujimotos @niedbalski @edsiper @koleini i have a doc for this plugin coming soon, in the meantime if i could get some eyes on this that would be appreciated 🙂

**Comment by victor88121** [comment]
Doc PR: https://github.com/fluent/fluent-bit-docs/pull/714

**Comment by agup006** [comment]
@nokute78 would you be able to help review?

**Pull request merged**